Cython/Compiler/Symtab.py

```diff
@@ -493,17 +493,28 @@ def declare(self, name, cname, type, pos, visibility, shadow = 0, is_type = 0, c
             warning(pos, "'%s' is a reserved name in C." % cname, -1)
 
         entries = self.entries
+        entry = None
+        cpp_already_overridden = False
         if name and name in entries and not shadow and not self.is_builtin_scope:
             old_entry = entries[name]
 
             # Reject redeclared C++ functions only if they have the same type signature.
             cpp_override_allowed = False
             if type.is_cfunction and old_entry.type.is_cfunction and self.is_cpp():
                 for alt_entry in old_entry.all_alternatives():
-                    if type == alt_entry.type:
+                    if type.compatible_signature_with(alt_entry.type):
                         if name == '<init>' and not type.args:
                             # Cython pre-declares the no-args constructor - allow later user definitions.
                             cpp_override_allowed = True
+                        elif alt_entry.is_inherited:
+                            cpp_override_allowed = True
+                        if cpp_override_allowed:
+                            alt_entry.type = type
```

Comment:
By "this line" I understand `alt_entry.type = type`. We can override methods which have a compatible signature in C++ (hence the change in line 505). But compatible doesn't mean exactly the same type.
So here, for the `alt_entry` which matches the redefinition, we update it's attributes to match the redefinition we are declaring, so that the updated entry looks exactly like if we didn't inherit anything.

Reply:
A comment for that will end up in the next batch then ;)